File: java-completed/app/src/main/java/com/vonage/tutorial/voice/view/oncall/OnCallViewModel.java
package com.vonage.tutorial.voice.view.oncall;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.nexmo.client.*;
import com.nexmo.client.request_listener.NexmoApiError;
import com.nexmo.client.request_listener.NexmoRequestListener;
import com.vonage.tutorial.R;
import com.vonage.tutorial.voice.util.CallManager;
import com.vonage.tutorial.voice.util.NavManager;
import timber.log.Timber;

public class OnCallViewModel extends ViewModel {
    private CallManager callManager = CallManager.getInstance();
    private NavManager navManager = NavManager.getInstance();

    private MutableLiveData<String> toastMutableLiveData = new MutableLiveData<>();
    LiveData<String> toastLiveData = toastMutableLiveData;

    private MutableLiveData<String> otherUserNameMutableLiveData = new MutableLiveData<>();
    LiveData<String> otherUserNameLiveData = otherUserNameMutableLiveData;

    public OnCallViewModel() {
        NexmoCall onGoingCall;

        if (callManager.getOnGoingCall() == null) {
            throw new RuntimeException("Call is null");
        } else {
            onGoingCall = callManager.getOnGoingCall();
        }

        onGoingCall.addCallEventListener(callEventListener);
    }

    private NexmoCallEventListener callEventListener = new NexmoCallEventListener() {
        @Override
        public void onMemberStatusUpdated(NexmoCallMemberStatus callMemberStatus, NexmoCallMember callMember) {
            Timber.d("CallEventListener.onMemberStatusUpdated: %s : %s", callMember.getUser().getName(), callMemberStatus);

            if (callMemberStatus == NexmoCallMemberStatus.COMPLETED || callMemberStatus == NexmoCallMemberStatus.CANCELED) {
                callManager.setOnGoingCall(null);
                navManager.popBackStack(R.id.mainFragment, false);
            }
        }

        @Override
        public void onMuteChanged(NexmoMediaActionState mediaActionState, NexmoCallMember callMember) {
            Timber.d("CallEventListener.onMuteChanged: %s : %s", callMember.getUser().getName(), mediaActionState);
        }

        @Override
        public void onEarmuffChanged(NexmoMediaActionState mediaActionState, NexmoCallMember callMember) {
            Timber.d("CallEventListener.onEarmuffChanged: %s : %s", callMember.getUser().getName(), mediaActionState);
        }

        @Override
        public void onDTMF(String dtmf, NexmoCallMember callMember) {
            Timber.d("CallEventListener.onDTMF: %s : %s", callMember.getUser().getName(), dtmf);
        }
    };

    @Override
    protected void onCleared() {
        super.onCleared();

        NexmoCall ongoingCall = callManager.getOnGoingCall();

        if (ongoingCall != null) {
            ongoingCall.removeCallEventListener(callEventListener);
        }
    }

    public void onBackPressed() {
        hangupInternal();
    }

    public void hangup() {
        hangupInternal();
    }

    private void hangupInternal() {
        NexmoCall ongoingCall = callManager.getOnGoingCall();

        if (ongoingCall != null) {
            ongoingCall.hangup(new NexmoRequestListener<NexmoCall>() {
                @Override
                public void onSuccess(@Nullable NexmoCall call) {
                    callManager.setOnGoingCall(null);
                }

                @Override
                public void onError(@NonNull NexmoApiError apiError) {
                    toastMutableLiveData.postValue(apiError.getMessage());
                }
            });
        }

    }

    public void onInit(OnCallFragmentArgs args) {
        otherUserNameMutableLiveData.postValue(args.getOtherUserName());
    }
}

File: kotlin-start/app/src/main/java/com/vonage/tutorial/voice/view/oncall/OnCallViewModel.kt
package com.vonage.tutorial.voice.view.oncall

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.nexmo.client.NexmoCall
import com.nexmo.client.NexmoCallEventListener
import com.nexmo.client.NexmoCallMember
import com.nexmo.client.NexmoCallMemberStatus
import com.nexmo.client.NexmoMediaActionState
import com.nexmo.client.request_listener.NexmoApiError
import com.nexmo.client.request_listener.NexmoRequestListener
import com.vonage.tutorial.R
import com.vonage.tutorial.voice.extension.asLiveData
import com.vonage.tutorial.voice.util.CallManager
import com.vonage.tutorial.voice.util.NavManager
import timber.log.Timber

class OnCallViewModel : ViewModel() {
    private val callManager = CallManager
    private val navManager = NavManager

    private val toastMutableLiveData = MutableLiveData<String>()
    val toastLiveData = toastMutableLiveData.asLiveData()

    private val otherUserNameMutableLiveData = MutableLiveData<String>()
    val otherUserNameLiveData = otherUserNameMutableLiveData.asLiveData()

    private val callEventListener = object : NexmoCallEventListener {
        override fun onMemberStatusUpdated(callMemberStatus: NexmoCallMemberStatus, callMember: NexmoCallMember) {
            Timber.d("CallEventListener.onMemberStatusUpdated: ${callMember.user.name} : $callMemberStatus")

            if (callMemberStatus == NexmoCallMemberStatus.COMPLETED || callMemberStatus == NexmoCallMemberStatus.CANCELED) {
                callManager.onGoingCall = null
                navManager.popBackStack(R.id.mainFragment, false)
            }
        }

        override fun onMuteChanged(mediaActionState: NexmoMediaActionState, callMember: NexmoCallMember) {
            Timber.d("CallEventListener.onMuteChanged: ${callMember.user.name} : $mediaActionState")
        }

        override fun onEarmuffChanged(mediaActionState: NexmoMediaActionState, callMember: NexmoCallMember) {
            Timber.d("CallEventListener.onEarmuffChanged: ${callMember.user.name} : $mediaActionState")
        }

        override fun onDTMF(dtmf: String, callMember: NexmoCallMember) {
            Timber.d("CallEventListener.onDTMF: ${callMember.user.name} : $dtmf")
        }
    }

    init {
        val onGoingCall = checkNotNull(callManager.onGoingCall) { "Call is null" }
        onGoingCall.addCallEventListener(callEventListener)
    }

    override fun onCleared() {
        super.onCleared()

        callManager.onGoingCall?.removeCallEventListener(callEventListener)
    }

    fun onBackPressed() {
        hangupInternal()
    }

    fun hangup() {
        hangupInternal()
    }

    private fun hangupInternal() {
        callManager.onGoingCall?.hangup(object : NexmoRequestListener<NexmoCall> {
            override fun onSuccess(call: NexmoCall?) {
                callManager.onGoingCall = null
            }

            override fun onError(apiError: NexmoApiError) {
                toastMutableLiveData.postValue(apiError.message)
            }
        })
    }

    fun onInit(args: OnCallFragmentArgs) {
        otherUserNameMutableLiveData.postValue(args.otherUserName)
    }
}
